feat: Add console scripts to start Deephaven from the command line

mofojed opened the pull request:
- Added console scripts to the embedded server to add a `deephaven` console script that can launch deephaven
- Tested by building and installing the `deephaven_server` whl
- `deephaven` is the root script, with `server` being the only positional "command" argument and ran by default
  - Figured I'd add command as a positional argument, in case we want to add other commands (e.g. `run` to run a specific file)
- Parse arguments for `host`, `port`, `jvm-args`, and `extra-classpath`
  - If no key provided, a random key is generated
- After starting, waits for the user to press Ctrl+C to exit

Examples:
`deephaven server`: Starts up Deephaven server on port 8080 with a random key and opens the web UI.
`deephaven server --port 9999`: Starts up Deephaven on port 9999.
`deephaven server --jvm-args="-Ddeephaven.dataDir=/tmp/deephaven"`: Starts up Deephaven with `/tmp/deephaven` as the data directory.
`deephaven server --port 9999 --jvm-args="-Ddeephaven.dataDir=/tmp/deephaven-data -Ddeephaven.console.disable=true"`: Starts up Deephaven on port 9999, using `/tmp/deephaven` as the data directory and the console disabled.
`deephaven --help`: Prints help about what commands are available
`deephaven server --help`: Prints help about the server comand

devinrsmith commented on py/embedded-server/deephaven_server/__init__.py:
```diff
@@ -14,3 +19,55 @@
+    if args.jvm_args is not None:
+        jvm_args += args.jvm_args.split()
+
+    s = Server(host=args.host, port=args.port, jvm_args=jvm_args, dh_args=args.dh_args)
```
Been a while since I've looked at this layer. It seems like `dh_args` doesn't do anything. We should probably remove it from the stack?

mofojed replied:
Yea they don't do anything, should I remove it from EmbeddedServer as well or just from here?

devinrsmith replied:
I think so. I would double check w/ @niloc132

devinrsmith commented on py/embedded-server/deephaven_server/__init__.py:
```diff
@@ -14,3 +19,55 @@
+    Args:
+        args: The arguments parsed using argparse to use to start the server.
+    """
+    if args.key is None:
+        args.key = secrets.token_urlsafe(32)
+
+    jvm_args = [f"-Dauthentication.psk={args.key}"]
+    if args.jvm_args is not None:
+        jvm_args += args.jvm_args.split()
```
Something we've talked about before is creating the config file if it doesn't exist and writing the PSK into it; this gives it some persistence. I think that's how jupyter does it?

mofojed replied:
Jupyter doesn't seem to persist the token. If I run `jupyter lab` twice in a row, I get new tokens. I think that makes sense, it shouldn't persist unless you specify it.

That being said I'm not against it and Jupyter certainly doesn't have to be precedence.

devinrsmith replied:
Okay; do you know if a fresh jupyter lab writes down any configuration files?

mofojed replied:
Yes, it writes a couple of files it looks like:
- `/home/bender/.jupyter/lab/workspaces/default-37a8.jupyterlab-workspace`: Config for the users workspace, such as layout of the panels
- `/home/bender/.local/share/jupyter/runtime/jupyter_cookie_secret`: I'm not 100% sure what this cookie is for; it's not the token used for auth (that regenerates each time) and it just regenerates this if you delete the file. Searching for details on what that cookie is used for isn't turning up much (I checked the [jupyter](https://github.com/search?q=org%3Ajupyter%20jupyter_cookie_secret&type=code) and [jupyterlab](https://github.com/search?q=org%3Ajupyterlab+jupyter_cookie_secret&type=code) orgs on GitHub). There's some mention of it in forums indicating it'll fail if [permissions don't allow reading the file](https://github.com/jupyterhub/jupyterhub/issues/1787), but it will regenerate if the file doesn't exist.

So again, we can write a config, I don't mind writing the token to a file for persistence, but that doesn't seem to be what Jupyter is storing (and I'm actually not sure _what_ the secret cookie they're storing is anyway. Some sort of base64 string, but it doesn't seem to be stored in the browser as a cookie, doesn't seem to have a reference in any of the accessible Jupyter code that I've searched ... I don't know what it is?

devinrsmith commented on py/embedded-server/deephaven_server/__init__.py:
```diff
@@ -14,3 +19,55 @@
+    Args:
+        args: The arguments parsed using argparse to use to start the server.
+    """
+    if args.key is None:
+        args.key = secrets.token_urlsafe(32)
+
+    jvm_args = [f"-Dauthentication.psk={args.key}"]
```
I tried to run locally, and these default jvm_args / host / port break my setup.
![Screenshot from 2024-03-21 19-08-22](https://github.com/deephaven/deephaven-core/assets/6764691/17d340fc-c57b-43ac-9687-14c778c73ea7)


Specifically, I run with:

```
$ cat ~/.config/deephaven/deephaven.prop                         
includefiles=dh-defaults.prop

http.port=8443
http.http1=false

ssl.identity.type=privatekey
ssl.identity.certChainPath=/home/devin/.config/deephaven/dev-certs/server.chain.crt
ssl.identity.privateKeyPath=/home/devin/.config/deephaven/dev-certs/server.key

AuthHandlers=io.deephaven.auth.AnonymousAuthenticationHandler
```

IMO, I think we either need to respect the default (or explicitly pointed to) config file (we can set up a proper way to query the config file in a proper way [ie, _not_ just reading it] so the client can get out the host, port); or, we need to say that deephaven-server is special and has its own config application namespace. That said, even if it _does_ have its own application namespace, there is still room for configuration files at that location that aren't compatible with this setup...

mofojed replied:
Yes that's a good point. I could get rid of the defaults and then it shouldn't conflict. However, the biggest issue I see is with the key. If I do not know the pre-shared key, I can't print the URL to which to open the UI for. I could still open the UI, but you'll be prompted to enter the key... which you won't know if you didn't specify it in your configuration file, which as a new user ad-hoc testing it out, is unlikely to have done.

So what's the best way to resolve it such that I can get the key that is configured? And/or get the key the server randomly generates?

devinrsmith replied:
We could probably hook up a generic way to get a URL from the server in process; this would allow it to _potentially_ work with other authorization methods if they chose to implement it. At a minimum, we'd implement it for anonymous and PSK. And it would also handle https cases. In my configuration above, the server would return `https://localhost:8443/ide/` since I'm running TLS + 8443 + anonymous.

mofojed replied:
I've changed it so it doesn't default a port, it allows the server startup to initialize it/read it from config.
The key I'm still generating, though also added a flag to run in `anonymous` mode... so in your case, it _should_ work for you if you run `deephaven --anonymous`.

> We could probably hook up a generic way to get a URL from the server in process

Sure, but that doesn't exist right now. Would you rather wait for that to get wired up and then change this to match? Or merge that down the road? Or change this PR in some other way?

chipkent commented on py/embedded-server/deephaven_server/__init__.py:
```diff
@@ -14,3 +19,55 @@
+    parser.add_argument("--dh-args", default=None, help="The Deephaven arguments to use.")
+    args = parser.parse_args()
+
+    if args.cmd == "start":
```
If you step back and think of where this script should go, "start" seems awkward.  It seems like you would want:
* Fresh IDE (--ide option which is default?)
* Fresh Python CLI (--no-ide option?)
* Run a script and then get an IDE (maybe you get this if you have a script ARG with the default --ide)
* Run batch script (maybe you get this if you have a script ARG and --no-ide)

mofojed replied:
Ah yea starting a console would be a good idea too. So:
```
deephaven                            # Aliases to `deephaven start`
deephaven start                    # Opens the Web UI in the browser
deephaven start --console    # Opens a python CLI instead of web UI
deephaven run <file>            # Run a specific file, can use the `--console` flag
```
If we default to `--ide`, doesn't make sense to have it as a flag (IMO), and then having a negative flag `--no-ide` seems odd.
Let me know what you think.

chipkent replied:
Look at how click does boolean flags, and my comment will make more sense.
https://click.palletsprojects.com/en/8.1.x/options/#boolean-flags

devinrsmith replied:
I think we could also consider having the `deephaven-core` wheel own the main `deephaven` entrypoint, as I imagine we _could_ use it in other contexts. When `deephaven-server` is in the environment, it could add-in `deephaven server` subcommand. I think some of these would be useful even if the user is running the native application workflow (ie, _not_ `deephaven-server`).

```sh
deephaven config show
deephaven plugins list
deephaven server start
```
etc

mofojed replied:
Yea that makes sense... it could also have the server start print out a message if `deephaven_server` isn't installed, something along the lines of:

```
def server():
  """
  Run a deephaven_server command
  """
  try:
    import deephaven_server
  except Exception as e:
    print("You need to install the `deephaven_server` package")
    sys.exit(1)
  ... # process the command
```

chipkent commented on py/embedded-server/deephaven_server/__init__.py:
```diff
@@ -14,3 +19,55 @@
+    Accepts a command to start the application and the arguments to use.
+    Defaults to `start`, which starts the server.
+    """
+    parser = argparse.ArgumentParser(description="Start the Deephaven embedded server")
+    parser.add_argument("cmd", nargs="?", default="start", choices=["start"], help="Deephaven command to run. Defaults to start.")
+    parser.add_argument("-p", "--port", default=8080, type=int, help="The port to bind to. Defaults to 8080.")
+    parser.add_argument("--host", default="localhost", help="The host to bind to. Defaults to localhost.")
+    parser.add_argument("--key", default=None, help="The key to use. Uses a randomly generated key if not specified.")
+    parser.add_argument("--jvm-args", default=None, help="The JVM arguments to use.")
+    parser.add_argument("--dh-args", default=None, help="The Deephaven arguments to use.")
+    args = parser.parse_args()
```
Click provides many useful features.  

Environment variables are just one important one.
https://click.palletsprojects.com/en/8.1.x/options/#values-from-environment-variables
https://click.palletsprojects.com/en/8.1.x/options/#multiple-values-from-environment-values

This could be combined with `.env` support to make a really flexible tool.
https://dev.to/jakewitcher/using-env-files-for-environment-variables-in-python-applications-55a1

mofojed replied:
I'll take a look

devinrsmith replied:
I don't have a horse in this race; just noting some stuff I learned recently. click is based on optparse, and optparse is soft deprecated. https://www.bitecode.dev/i/135568834/pep-defines-soft-deprecation-getopt-and-optparse-soft-deprecated

This article does specifically point out https://typer.tiangolo.com/ and https://click.palletsprojects.com/en/8.1.x/, so I think it's reasonable to compare argparse vs click vs typer.

mofojed replied:
Yea the nice thing about `argparse` is it's baked in. We're really not doing a lot here, I don't think it really matters which one we use?

chipkent replied:
It is going to matter what we use because the complexity of this is going to grow.  We don't want to break users because we outgrow `argparse`.

`typer` is built using `click`.  https://typer.tiangolo.com/alternatives/#click

mofojed replied:
Switched to using click.

devinrsmith commented:
Note from community user: 
> we noticed the deephaven_server python package doesn't allow for specifying extra class pass paths

We might want to consider inheriting the EXTRA_CLASSPATH environment variable (and or START_OPTS / JAVA_OPTS)?

devinrsmith reviewed (changes requested):
I've got some suggestions that expose more of the configuration properties to EmbeddedServer, PR https://github.com/mofojed/deephaven-core/pull/8.

The one other thing that I think will be important is moving the console entry point _out_ of `__init__.py`. I think users who import deephaven_server should not need to source the click / cli logic.

chipkent commented on py/embedded-server/deephaven_server/server.py:
```diff
@@ -7,10 +7,39 @@
 # before the JVM was running.
 
 
+class ServerConfig:
```
pydocs

chipkent commented on py/embedded-server/deephaven_server/server.py:
```diff
@@ -7,10 +7,39 @@
+        return self.j_server_config
+
+    @property
+    def target_url_or_default(self) -> str:
```
pydocs

chipkent commented on py/embedded-server/deephaven_server/server.py:
```diff
@@ -7,10 +7,39 @@
+    def target_url_or_default(self) -> str:
+        return self.j_server_config.targetUrlOrDefault()
+
+    def __init__(self, j_server_config):
```
`__init__` is traditionally the first method in a class.

chipkent commented on py/embedded-server/deephaven_server/server.py:
```diff
@@ -7,10 +7,39 @@
+    def urls(self, target_url: str) -> List[str]:
+        return list(self.j_authentication_handler.urls(target_url).toArray())
+
+    def __init__(self, j_authentication_handler):
```
`__init__` is traditionally the first method in a class.

chipkent commented on py/embedded-server/deephaven_server/server.py:
```diff
@@ -7,10 +7,39 @@
+        self.j_server_config = j_server_config
+
+
+class AuthenticationHandler:
```
pydocs

chipkent commented on py/embedded-server/deephaven_server/server.py:
```diff
@@ -7,10 +7,39 @@
+        return self.j_authentication_handler
+
+    @property
+    def auth_type(self) -> str:
```
pydocs

chipkent commented on py/embedded-server/deephaven_server/server.py:
```diff
@@ -7,10 +7,39 @@
+    def auth_type(self) -> str:
+        return self.j_authentication_handler.getAuthType()
+
+    def urls(self, target_url: str) -> List[str]:
```
pydocs

chipkent commented on py/embedded-server/deephaven_server/server.py:
```diff
@@ -23,16 +52,28 @@ def j_object(self):
 
     @property
-    def port(self):
+    def port(self) -> int:
```
pydocs

chipkent commented on py/embedded-server/deephaven_server/server.py:
```diff
@@ -23,16 +52,28 @@ def j_object(self):
         return self.j_server.getPort()
 
+    @property
+    def server_config(self) -> ServerConfig:
```
pydocs

chipkent commented on py/embedded-server/deephaven_server/server.py:
```diff
@@ -23,16 +52,28 @@ def j_object(self):
+        return ServerConfig(self.j_server.serverConfig())
+
+    @property
+    def authentication_handlers(self) -> List[AuthenticationHandler]:
```
pydocs

chipkent commented on py/embedded-server/deephaven_server/__init__.py:
```diff
@@ -3,14 +3,93 @@
+        sys.exit(0)
+
+    signal.signal(signal.SIGINT, signal_handler)
+    signal.pause()
```
This is looking great.

1. I would add an optional argument to `server` that takes a path to a script file to run before getting into the wait loop.  Basically a way to populate DH with a quick command.
2. I would add a new `run` command to run a script.  Basically the same as `server` without the auth, browser, and wait.  Batch mode.

mofojed replied:
Both of those options sound great. Mind if it's a future PR?:
1. That's kind of like application mode, would we want to specify the "app" to start with? I believe the only way to do that now is by setting the app.d directory, so we'd want to think how we have this play out/work with application mode.
2. Similar to question 1, do we still have the application mode scripts run or just the file you specified? What config do we want to override in that case?

In any case with the console scripts added, we can add new commands/options fairly easily.

chipkent replied:
Let's do it soon.

niloc132 commented on py/embedded-server/deephaven_server/__init__.py:
```diff
@@ -3,14 +3,93 @@
+    if anonymous:
+        jvm_args += [f"-DAuthHandlers=io.deephaven.auth.AnonymousAuthenticationHandler"]
+    elif key is not None:
+        jvm_args += [
```
Looks like this prevents a user from specifying mTLS/etc as their preferred option, without forcing that they also have PSK (or anon) enabled as well... and it will override their `AuthHandlers` value and wipe out their own config.

Unsure if we want to get into the business of parsing AuthHandlers or other config vars here... if we want to go after this it might be time to tighten up our config language/control, or just be prepared for breaking changes.

Or put dh_args back or something, and require that it be used instead of system props for dh configs? and have some kind of schema here for "this is actually a `,`-delimited list of strings"?

I kind of hate all of the choices... I do have a draft proposal for "we should use TOML", but I'm not sure now's the time to pick that up in earnest?

niloc132 replied:
Additionally, these shouldn't be mutually exclusive - one day we'll have authorization hooks that are easy to use, and we'll want to allow both (anon might be simple, psk might be admin).

devinrsmith replied:
I'm _hoping_ we can discourage users from setting `--anonymous` or `--psk` wrt a deephaven server start command (in my helper PR I originally removed them). When they are not provided, the server will retain it's normal defaults or source from the users configuration file. I also suggested that `deephaven set-config ...` might be a reasonable command that allows us to explicitly present a way to do "proper" configuration from the CLI (ie, writing down into a configuration file w/ proper safeguards).

There was a similar concern I had for jsPlugins configuration, essentially relying on a prefix call:

```java
class PropertyFile {
...
    /**
     * Collect all of the properties in this property file that begin with a given prefix.
     *
     * @return a new Properties object containing the selected properties, with the prefix removed.
     */
    public Properties getProperties(String prefix) {
```

to allow independent configuration. ie

```properties
deephaven.jsPlugins.foo=/path/to/plugin-1
deephaven.jsPlugins.bar=/path/to/plugin-2
```

as opposed to needing to do

```properties
deephaven.jsPlugins=/path/to/plugin-1,/path/to/plugin-2
```

devinrsmith replied:
I don't think we want or need `deephaven server` command line options to present the full universe of configuration possibilities to users; in my mind, these options may be better thought of as `--only-psk=<key>`, `--only-anonymous`, etc. I am leaning that maybe we should remove them from `deephaven server` altogether, I don't think it would be a big loss. (I could make similar argument for `--host` and `--port`, but I think that's a slightly harder sell.)

mofojed replied:
I'd be alright with getting rid of the anonymous/psk options now that there's the `targetUrls` option, as those options conflicted with existing config. Leaving `jvm_args` seems alright, it adds to the args in your existing config. `host`/`port` seem like they'd be nice and would just override any config you have which makes sense IMO.

mofojed commented:
Added all the missing pydocs and javadocs. Deprecated the dh_args parameter from EmbeddedServer - I didn't remove it entirely.

I left the `--key` and `--anonymous` options - they are convenient, and it would be opt-in - by default it'll still use whatever your config has defined for auth.

@devinrsmith I also moved it out of `__init__.py` to `cli.py`. Let me know if it should be somewhere else entirely; notably, the `click` is still in `install_requires` for `deephaven_server`.

mofojed commented on py/embedded-server/deephaven_server/cli.py:
```diff
@@ -0,0 +1,79 @@
+    s = Server(host=host, port=port, jvm_args=jvm_args)
+    s.start()
+
+    target_url_or_default = s.server_config.target_url_or_default
+    authentication_type = None
+    authentication_url = target_url_or_default
+    for authentication_handler in s.authentication_handlers:
+        authentication_urls = authentication_handler.urls(target_url_or_default)
+        if authentication_urls:
+            authentication_type = authentication_handler.auth_type
+            authentication_url = authentication_urls[0]
+            break
```
@devinrsmith / @niloc132 any reason we don't just have a method on `EmbeddedServer` that does all this song and dance to return all the auth URLs?
I guess we could have multiple authentication handlers, but I feel like 90% of the time it's just going to be one auth handler and we'll have one default URL. Doesn't matter much though.

devinrsmith replied:
That's the route I originally took when creating this feature; but it became clear that I was trying to service this one specific use case as opposed to being more library-like and presenting the components themselves.

I could see an argument for moving this logic into the _python_ side `Server`; I don't necessarily think the java side `EmbeddedServer` is the correct place. Once you elevate it to `Server` though, it becomes a public API and something we need to treat as a library feature more generally.

niloc132 replied:
With that said, this cli _is_ a public API - once we put it out there, we can't break it (without being the kind of jerk that would break any other API).

mofojed commented on py/embedded-server/deephaven_server/server.py:
```diff
@@ -7,10 +7,69 @@
+        """
+        return self.j_authentication_handler.getAuthType()
+
+    def urls(self, target_url: str) -> List[str]:
```
@devinrsmith Also this gives the URL to the IDE - just wondering, if we want URLs to the iframe endpoints as well, that could be useful for deephaven-ipywidgets (and others).

devinrsmith replied:
This API doesn't assume much about `target_url`; in the case of anonymous, it just echos it back, and in the case of PSK, just appends `?psk=<key>`. It doesn't know anything about IDE nor iframe, etc.

By default, `target_url = "http://localhost:10000"`. As part of CLI, you _could_ append `/ide/` to be a bit more explicit, or invoke the logic multiple times to get iframe or other links.

It would be nice to have a more proper API around web apps installed as opposed to doing:

```java
for (String appRoot : List.of("/ide/", "/iframe/table/", "/iframe/chart/", "/iframe/widget/")) {
```

Until then though, I think you'll need to do something similar in CLI if you want to.

niloc132 commented on py/embedded-server/deephaven_server/cli.py:
```diff
@@ -0,0 +1,82 @@
+@click.option(
+    "--key",
+    default=None,
+    help="Start the server in pre-shared key mode. If the key is empty, a random key will be used. Cannot be used if anonymous is specified.",
```
> Cannot be used if anonymous is specified.

Why not? Aside from limitations of the patch, this is a reasonable thing to do and supported by the server? The authorization APIs would need to be used in some way to really be helpful ("anon users can access everything except the console, psk users can access everything"), but none of these are mutually exclusive.

mofojed replied:
Removed this option.

jmao-denver commented on py/embedded-server/deephaven_server/cli.py:
```diff
@@ -0,0 +1,59 @@
+    pass
+
+
+@cli.command()
+@click.option("--host", default=None, help="The host to bind to.")
+@click.option("--port", default=None, type=int, help="The port to bind to.")
+@click.option("--jvm-args", default=None, help="The JVM arguments to use.")
+def server(host, port, jvm_args):
+    """
+    Start the Deephaven server.
+    """
+    click.echo("Starting Deephaven server...")
+
+    if jvm_args is None:
+        jvm_args = ""
+    jvm_args = jvm_args.split()
+
+    s = Server(host=host, port=port, jvm_args=jvm_args)
+    s.start()
```
Is there any good reason to not support the `extra_classpath` argument recently added to `Server()`?

mofojed replied:
That was added while this PR was in flight. I don't see why not, and I think @devinrsmith had that intention.

mofojed replied:
@jmao-denver added

jmao-denver reviewed (approved):
LGTM

deephaven-internal commented:
Labels indicate documentation is required. Issues for documentation have been opened:

Community: https://github.com/deephaven/deephaven-docs-community/issues/185
